### PR TITLE
Make changes to support compiling on Bookworm (with GCC 12)

### PR DESCRIPTION
**Pull request by saiarcot895**
This adds support for sairedis to compile on Debian Bookworm (with GCC 12). This PR also includes a SAI submodule update that brings in other unrelated changes; this is needed because of a change in Doxyfile that would cause warnings/errors in newer versions of Doxygen.

The SAI submodule update brings in the following changes:
* Update the Doxyfile for doxygen in Debian Bookworm (opencomputeproject/SAI#1946)
* Enable sai_uint16_t in ProcessStructValueType Struct Member (opencomputeproject/SAI#1949)
* [meta] Add support for port stat extensions (opencomputeproject/SAI#1947)
* [meta] Add custom range start end values check (opencomputeproject/SAI#1945)
* Cable diagnostics attribute added (opencomputeproject/SAI#1894)
* Add attributes to disable L3 rewrites (opencomputeproject/SAI#1924)
* Add MAC remote loopback to the port loopback enums. (opencomputeproject/SAI#1934)
* [TAM] Granular counter subscription (opencomputeproject/SAI#1670)

Note that runtime functionality testing has not been done; only compilation (with the build-time tests) has been done.

### CODE REVIEW

File: syncd/ServiceMethodTable.cpp [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 #include "swss/logger.h"
 
+#include <array>
```
Comment (kcudnik) [inline syncd/ServiceMethodTable.cpp:5]: Why this is added?

File: syncd/SwitchNotifications.cpp [hunk 2]
```diff
@@ -1,6 +1,7 @@
 #include "SwitchNotifications.h"
 
 #include "swss/logger.h"
+#include <array>
```
Comment (kcudnik) [inline syncd/SwitchNotifications.cpp:4]: Why this isnadded?

**Review by kcudnik — APPROVED** [review]

File: syncd/SwitchNotifications.cpp (hunk 2 above)
Comment (saiarcot895) [reply syncd/SwitchNotifications.cpp:4]: GCC 12 complained that `std::array` was not completely defined, but is instantiated on line 144.

Snippet of error log:

```
SwitchNotifications.cpp: In instantiation of 'constexpr auto declare_static(std::index_sequence<i ...>) [with B = syncd::SwitchNotifications::SlotBase; D = syncd::SwitchNotifications::Slot; long unsigned int ...i = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15}; std::index_sequence<i ...> = std::integer_sequence<long unsigned int, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15>]':
SwitchNotifications.cpp:151:35:   required from 'constexpr auto declare_static() [with B = syncd::SwitchNotifications::SlotBase; D = syncd::SwitchNotifications::Slot; long unsigned int size = 16]'
SwitchNotifications.cpp:156:83:   required from here
SwitchNotifications.cpp:144:56: error: invalid use of incomplete type 'struct std::array<syncd::SwitchNotifications::SlotBase*, 16>'
  144 |     return std::array<B*, sizeof...(i)>{{new D<i>()...}};
      |                                                        ^
In file included from /usr/include/c++/12/bits/stl_map.h:63,
                 from /usr/include/c++/12/map:61,
                 from /usr/include/swss/logger.h:6,
                 from SwitchNotifications.h:7,
                 from SwitchNotifications.cpp:1:
/usr/include/c++/12/tuple:1595:45: note: declaration of 'struct std::array<syncd::SwitchNotifications::SlotBase*, 16>'
...
```

File: syncd/ServiceMethodTable.cpp (hunk 1 above)
Comment (saiarcot895) [reply syncd/ServiceMethodTable.cpp:5]: GCC 12 complained that `std::array` was not completely defined, but is instantiated on line 74.

Snippet of error log:

```
ServiceMethodTable.cpp: In instantiation of 'constexpr auto declare_static(std::index_sequence<i ...>) [with B = syncd::ServiceMethodTable::SlotBase; D = syncd::ServiceMethodTable::Slot; long unsigned int ...i = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9}; std::index_sequence<i ...> = std::integer_sequence<long unsigned int, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9>]':
ServiceMethodTable.cpp:81:35:   required from 'constexpr auto declare_static() [with B = syncd::ServiceMethodTable::SlotBase; D = syncd::ServiceMethodTable::Slot; long unsigned int size = 10]'
ServiceMethodTable.cpp:86:79:   required from here
ServiceMethodTable.cpp:74:56: error: invalid use of incomplete type 'struct std::array<syncd::ServiceMethodTable::SlotBase*, 10>'
   74 |     return std::array<B*, sizeof...(i)>{{new D<i>()...}};
      |                                                        ^
In file included from /usr/include/c++/12/bits/stl_map.h:63,
                 from /usr/include/c++/12/map:61,
                 from /usr/include/swss/logger.h:6,
                 from ServiceMethodTable.h:7,
                 from ServiceMethodTable.cpp:1:
/usr/include/c++/12/tuple:1595:45: note: declaration of 'struct std::array<syncd::ServiceMethodTable::SlotBase*, 10>'
 1595 |   template<typename _Tp, size_t _Nm> struct array;
      |                                             ^~~~~
ServiceMethodTable.cpp: In instantiation of 'constexpr auto declare_static() [with B = syncd::ServiceMethodTable::SlotBase; D = syncd::ServiceMethodTable::Slot; long unsigned int size = 10]':
ServiceMethodTable.cpp:86:79:   required from here
ServiceMethodTable.cpp:82:48: error: no matching function for call to 'std::vector<syncd::ServiceMethodTable::SlotBase*>::vector(<brace-enclosed initializer list>)'
   82 |     return std::vector<B*>{arr.begin(), arr.end()};
      |                                         ~~~~~~~^~
```

Comment (kcudnik) [reply syncd/ServiceMethodTable.cpp:5]: 3hivh version cimpiler

Comment (saiarcot895) [reply syncd/ServiceMethodTable.cpp:5]: GCC 12.2, specifically (also using libstdc++ 12.2)

Comment (saiarcot895) [reply syncd/ServiceMethodTable.cpp:5]: If you want, I think I can modify the pipeline to build this on Bookworm as well, so that is covered.

Comment (kcudnik) [reply syncd/ServiceMethodTable.cpp:5]: Your title suggest tat this is the change so support bookworm so what other changes would be needed ?

Comment (saiarcot895) [reply syncd/ServiceMethodTable.cpp:5]: Not sure at this point. At this point, only compilation (and the tests that get run as part of the build) have been done; runtime testing has not been done yet.

Comment (kcudnik) [reply syncd/ServiceMethodTable.cpp:5]: Then please update title to reflect changes, like GCC latest support or update headers etc

Comment (saiarcot895) [reply syncd/ServiceMethodTable.cpp:5]: Updated

**Pull request merged**